### PR TITLE
EDR - Event Data Records

### PR DESCRIPTION
**Pull request by kalda341**
http://opentelecom.foundation/pull-requests/kazoo/contributions/2017/08/10/kazoo-edr-event-data-records.html

### CODE REVIEW

File: applications/edr/src/edr.app.src [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  {applications, [ kernel
+                 , stdlib
+
+                 , kazoo_apps
+                 , kazoo_auth
+                 , kazoo_edr
+
+                 , lager
+                 , ranch
+                 , cowboy
```
Comment (jamesaimonetti) [inline applications/edr/src/edr.app.src]: Is edr listening on an http socket?

File: core/kazoo/src/kz_util.erl [hunk 2]
```diff
@@ -591,6 +592,15 @@ account_update(Account, UpdateFun) ->
             account_update(UpdateFun(AccountJObj))
     end.
 
+-spec account_tree(api_binary()) -> api_binaries() | 'undefined'.
```
Comment (jamesaimonetti) [inline core/kazoo/src/kz_util.erl]: Please put this in `kz_account` and have it do the fetch/get in account_tree/1

**Comment by jamesaimonetti** [comment]
@kalda341 You'll need to rebase against master. Also, consider running at least `./scripts/dialyze_changed.bash` to see the plethora of complaints Dialyzer has. I believe with Erlang 19 you don't get the implicit 'undefined' in the record specs anymore so you'll need to update the specs to account for fields that are unset at any given time.

Also, I would recommend using `api_ne_binary()` instead of `api_binary()` unless you are sure the variable could be `<<>>`.

I would focus first on the `Record construction...` errors first as those are typically just updating the specs of the record in question. You'll see a slew of `Function XYZ has no local return` but typically at the end of a bunch of those will be the actual issue, something like `The call M:F(Args...) will never return since it differs in the Nth argument...`.

Once you've gotten most of those warnings cleaned up, you can add additional beams to let Dialyzer give you more feedback. I would at least add kazoo_stdlib and gen_listener:

```bash
./scripts/dialyze_changed.bash core/kazoo_stdlib/ebin core_kazoo_amqp/ebin/gen_listener.beam
```
You can add `ebin/` or individual `.beam` files to improve Dialyzer's information. I use the `Unknown functions:` listing to give me an idea of what modules/applications I should add to the command.

You will need to generate the initial PLT if you haven't already: `make build-plt` should do it. Recommend using 19.3 to compile and build the PLT with.

File: applications/edr/doc/README.md [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+## Backends
+
+The EDR app is essentially a collection of backends. Each backend will recieve EDR events matching
+a binding, and send them somewhere else (usually out of Kazoo).
+
+Backends are registered in system_config/edr.
+
+### General registerd backend configuration options
+
+#### name
```
[the remark this answers is not in the document]
Comment (kalda341) [reply applications/edr/doc/README.md]: @jamesaimonetti  There is a schema for this: https://github.com/2600hz/kazoo/pull/4194/files#diff-6becd372021cbc35a0a1b653d6da5af8 . Is there an easy way to generate this from the schema?

File: applications/edr/src/edr.app.src (hunk 1 above)
Comment (kalda341) [reply applications/edr/src/edr.app.src]: No, good catch!

File: applications/edr/doc/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jamesaimonetti) [reply applications/edr/doc/README.md]: @kalda341 in the VM, try:
```erlang
1> io:format("~s~n", [kz_ast_util:schema_to_table(<<"system_config.edr">>)]).
```
Should print out the copy/paste-able markdown table

**Comment by kalda341** [comment]
Thanks @jamesaimonetti  for the great feedback!

**Comment by jamesaimonetti** [comment]
@kalda341 i think this needs a rebase. Also, @noahmehl any update on testing?

**Comment by noahmehl** [comment]
@jamesaimonetti no update on testing yet :(

**Comment by jamesaimonetti** [comment]
@kalda341 @noahmehl how is this going? Looks like a rebase is in order again (conflicts now with master). Anything we can do to help move this along?

**Comment by kalda341** [comment]
@jamesaimonetti We have a test cluster set up, we just need to figure out a way to get some benchmarks with EDR off and on. Any suggestions would be greatly appreciated.
I'll update to master.

**Comment by kjcsb1** [comment]
We have configured 6 servers as follows: 
Kazoo: c-01-kz-01.opentelecom.foundation
AMQP: c-01-amqp-01.opentelecom.foundation
Freeswitch: c-01-fs-01.opentelecom.foundation
DB: c-01-db-01.opentelecom.foundation
Kamailio: c-01-km-01.opentelecom.foundation
Load test server: c-01-load-01.opentelecom.foundation

We then installed latest Kazoo (4.2.11), enabled ACDC, set a Zabbix agent on each server, and set "calls_consume_tokens" to false.

For testing we've created 10 accounts with a number pointed at a queue each. We then place 5 calls per second in to each number to a maximum of 50 calls per queue i.e. max 500 calls. We used sipp with the following scenario.
`<?xml version="1.0"?>
<scenario name="csb_test_20180130_01">
  <send retrans="500">
<![CDATA[
INVITE sip:[service]@[remote_ip]:[remote_port] SIP/2.0
Via: SIP/2.0/[transport] [local_ip]:[local_port];branch=[branch]
From: "sipp" <sip:sipp@[local_ip]:[local_port]>;tag=[call_number]
To: <sip:[service]@[remote_ip]:[remote_port]>
Call-ID: [call_id]
CSeq: [cseq] INVITE
Contact: <sip:sipp@[local_ip]:[local_port];transport=[transport]>
Max-Forwards: 100
User-Agent: SIPp/sippy_cup
Content-Type: application/sdp
Content-Length: [len]

v=0
o=user1 53655765 2353687637 IN IP[local_ip_type] [local_ip]
s=-
c=IN IP[media_ip_type] [media_ip]
t=0 0
m=audio [media_port] RTP/AVP 0 101
a=rtpmap:0 PCMU/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-15
]]>
<action><assignstr assign_to="remote_addr" value="[service]@[remote_ip]:[remote_port]"/><assignstr assign_to="local_addr" value="sipp@[local_ip]:[local_port]"/><assignstr assign_to="call_addr" value="[service]@[remote_ip]:[remote_port]"/></action></send>
  <recv response="100" optional="true"/>
  <recv response="180" optional="true"/>
  <recv response="183" optional="true"/>
  <recv response="200" rrs="true" rtd="true">
    <action>
      <ereg regexp="&lt;sip:(.*)&gt;.*;tag=([^;]*)" search_in="hdr" header="To:" assign_to="dummy,remote_addr,remote_tag"/>
    </action>
  </recv>
  <send>
<![CDATA[
ACK [next_url] SIP/2.0
Via: SIP/2.0/[transport] [local_ip]:[local_port];branch=[branch]
From: "sipp" <sip:sipp@[local_ip]:[local_port]>;tag=[call_number]
To: <sip:[service]@[remote_ip]:[remote_port]>[peer_tag_param]
Call-ID: [call_id]
CSeq: [cseq] ACK
Contact: <sip:[$local_addr];transport=[transport]>
Max-Forwards: 100
User-Agent: SIPp/sippy_cup
Content-Length: 0
[routes]
]]>
</send>
  <nop>
    <action>
      <exec play_pcap_audio="/root/csb_test_20180130_01.pcap"/>
    </action>
  </nop>
  <send>
<![CDATA[
ACK [next_url] SIP/2.0
Via: SIP/2.0/[transport] [local_ip]:[local_port];branch=[branch]
From: "sipp" <sip:sipp@[local_ip]:[local_port]>;tag=[call_number]
To: <sip:[service]@[remote_ip]:[remote_port]>[peer_tag_param]
Call-ID: [call_id]
CSeq: [cseq] ACK
Contact: <sip:[$local_addr];transport=[transport]>
Max-Forwards: 100
User-Agent: SIPp/sippy_cup
Content-Length: 0
[routes]
]]>
</send>
  <nop>
    <action>
      <exec play_pcap_audio="/root/csb_test_20180130_01.pcap"/>
    </action>
  </nop>
  <pause milliseconds="20000"/>
  <send>
<![CDATA[
BYE sip:[$call_addr] SIP/2.0
Via: SIP/2.0/[transport] [local_ip]:[local_port];branch=[branch]
From: <sip:[$local_addr]>;tag=[call_number]
To: <sip:[$remote_addr]>;tag=[$remote_tag]
Contact: <sip:[$local_addr];transport=[transport]>
Call-ID: [call_id]
CSeq: [cseq] BYE
Max-Forwards: 100
User-Agent: SIPp/sippy_cup
Content-Length: 0
[routes]
]]>
</send>
  <recv response="200"/>
  <Reference variables="remote_addr,local_addr,call_addr,dummy,remote_tag"/>
</scenario>
`

We flushed the database cache
sup kazoo_data_maintenance flush
date
Thu Feb  1 21:41:34 UTC 2018

Then ran the test
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000019 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000020 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000021 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000022 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000023 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000024 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000025 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000026 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000027 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000028 -r 2 -l 100 -m 50 -bg

This created a baseline test. 

We then installed EDR (but didn't enable it) and ran the same test. 
sup kazoo_data_maintenance flush
date
Thu Feb  1 21:41:34 UTC 2018
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000019 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000020 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000021 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000022 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000023 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000024 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000025 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000026 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000027 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000028 -r 2 -l 100 -m 50 -bg

We then enabled EDR and ran the same test. 
EDR will publish an event for each call that enters the queue.

Setup edr to have an http and amqp backend
system_configs/edr
`{
    "data": {
        "default": {
            "backends": [
	       {
           "name": "request_bin",
           "type": "http",
           "options": {
               "url": "https://requestb.in/1lldmi11",
               "headers": {
                   "Content-Type": "application/json"
               }
           },
           "bindings": [
               {
                   "account_id": "*",
                   "include_descendants": false,
                   "app_name": "*",
                   "severity": "ok",
                   "exact_severity": false,
                   "verbosity": "info",
                   "exact_verbosity": false
               }
           ],
           "enabled": true
       },
                  {
           "name": "amqp",
           "type": "amqp",
           "options": {
           },
           "bindings": [
               {
                   "account_id": "*",
                   "include_descendants": false,
                   "app_name": "*",
                   "severity": "ok",
                   "exact_severity": false,
                   "verbosity": "info",
                   "exact_verbosity": false
               }
           ],

             "enabled": true
       }
       ]
        },
        "id": "edr"
    }
}
`

Start edr
sup kapps_controller start_app edr

sup kazoo_data_maintenance flush
date
Fri Feb  2 00:55:28 UTC 2018

sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000019 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000020 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000021 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000022 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000023 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000024 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000025 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000026 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000027 -r 2 -l 100 -m 50 -bg
sipp -sf /root/csb_test_20180130_01.xml 172.31.31.169 -s 6430000028 -r 2 -l 100 -m 50 -bg

We verified that edr was pushing http data to request bin. We measured CPU and memory utilisation on the Kazoo and RabbitMQ servers

Results
Test 1 21:49
KZ Memory Available 7GB
KZ CPU Idle 30%
AMQP Memory Available 7GB
AMQP CPU Idle 97%
Test 2 23:29
KZ Memory Available 6.1GB
KZ CPU Idle 67%
AMQP Memory Available 7GB
AMQP CPU Idle 82%
Test 3 00:55
KZ Memory Available 6.1GB
KZ CPU Idle 61%
AMQP Memory Available 6GB
AMQP CPU Idle 83%

It doesn't appear that edr has a major impact on performance. 

Any thoughts or comments welcome.

**Comment by k-anderson** [comment]
@kalda341 Howdy, this is excellent thank you!

So there are some minor CI issues as a result of the recent changes that I have addressed on #4542 (commit https://github.com/2600hz/kazoo/pull/4542/commits/8d24dda4fd432c2e5504218309753fc5d43c37d8).  

I opened the PR so the tests would run but I am not planning on merging that PR, just checking the changes.  If you are available to cherry-pick that commit into this branch (assuming the tests pass) I think this PR is ready!  

If you don't have the ability to do so I could also manually merge this adding the commit as I do, just need your ok on the changes.

Thanks!

**Comment by kalda341** [comment]
@k-anderson I've cherry picked your changes. Thanks!

**Pull request merged**